File: chap19/Contraste.py
# Ces programmes sont sous licence CeCILL-B V1.
f1 = open("joconde.ppm","r")
   
s = f1.readline()
s = f1.readline()
largeur, hauteur = map( int, f1.readline().split() )
max = int(f1.readline())

rouge = [[255]*hauteur for loop in range(largeur)]
vert  = [[255]*hauteur for loop in range(largeur)]
bleu  = [[255]*hauteur for loop in range(largeur)]

for j in range(hauteur):
   for i in range(largeur):
      rouge[i][j] = int(f1.readline())
      vert [i][j] = int(f1.readline())
      bleu [i][j] = int(f1.readline())
f1.close()

t = [[255]*hauteur for loop in range(largeur)]
for i in range(largeur):
   for j in range(hauteur):
      t[i][j] = (rouge[i][j] + bleu[i][j] + vert[i][j]) // 3

t1 = [[255]*hauteur for loop in range(largeur)]
for i in range(largeur):
   for j in range(hauteur):
      if t[i][j] <= max // 5:
         t1[i][j] = 0
      else:
         t1[i][j] = max

f2 = open("jocondenoiretblanc.pgm","w")
print("P2",file=f2)
print("#",file=f2)
print(largeur,hauteur,file=f2)
print(max,file=f2)
for j in range(hauteur):
   for i in range(largeur):
      print(t1[i][j],file=f2)
f2.close()

File: chap19/Contraste.cpp
// Ces programmes sont sous licence CeCILL-B V1.
#include <iostream>
#include <fstream>
#include <string>

using namespace std;

const int TAILLE_MAX = 1000;
int r[TAILLE_MAX][TAILLE_MAX];
int v[TAILLE_MAX][TAILLE_MAX];
int b[TAILLE_MAX][TAILLE_MAX];
int t[TAILLE_MAX][TAILLE_MAX];
int t1[TAILLE_MAX][TAILLE_MAX];

int main() 
{
   int i,j,l,h,max; 
   string s;
   ifstream f1;  
   ofstream f2;  

   f1.open("joconde.ppm");
   getline(f1,s);
   getline(f1,s);
   f1 >> l;
   f1 >> h;
   f1 >> max;
   for (j = 0; j <= h-1; j = j + 1) 
   {
      for (i = 0; i <= l-1; i = i + 1)
      {
         f1 >> r[i][j];
         f1 >> v[i][j];
         f1 >> b[i][j];
      }
   }
   f1.close();

   for (i = 0; i <= l-1; i = i + 1)
   {
      for (j = 0; j <= h-1; j = j + 1)
      {
         t[i][j] = (r[i][j] + b[i][j] + v[i][j])/3;
      }
   }

   for (i = 0; i <= l-1; i = i + 1) 
   {
      for (j = 0; j <= h-1; j = j + 1)
      {
         if (t[i][j] <= max/5) 
         {
            t1[i][j] = 0;
         } 
         else 
         {
            t1[i][j] = max;
         }
      }
   }

   f2.open("jocondenoiretblanc.pgm");
   f2 << "P2" << endl;
   f2 << "#" << endl;
   f2 << l;
   f2 << " ";
   f2 << h << endl;
   f2 << max << endl;
   for (j = 0; j <= h-1; j = j + 1)
   {
      for (i = 0; i <= l-1; i = i + 1) 
      {
         f2 << t1[i][j] << endl;
      }
   }
   f2.close();
}
